signer/src/block_notifier.rs

```diff
@@ -0,0 +1,150 @@
+pub trait BlockNotifier {
+    /// Errors occurring during subscription or running the notifier.
+    type Error: Error;
+
+    /// Returns a stream of block headers.
+    fn subscribe(&self) -> Pin<Box<dyn Stream<Item = Result<BlockHash, Self::Error>> + Send>>;
+}
+
+/// A struct implementing the `BlockNotifier` trait using Electrum client.
+pub struct ElectrumBlockNotifier {
+    client: Arc<Client>,
+    retry_interval: Duration,
+    max_retry_attempts: u32,
+    ping_interval: Duration,
+    subscribe_interval: Duration,
+}
+
+impl ElectrumBlockNotifier {
+    /// Creates a new instance of `ElectrumBlockNotifier` from config settings.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `ElectrumBlockNotifier`.
+    pub fn from_config() -> Result<Self, ElectrumError> {
+        let server = &SETTINGS.block_notifier.server;
+        let config = ConfigBuilder::new().build();
+        let client = Arc::new(Client::from_config(server, config)?);
+        let retry_interval = Duration::from_secs(SETTINGS.block_notifier.retry_interval);
+        let max_retry_attempts = SETTINGS.block_notifier.max_retry_attempts;
+        let ping_interval = Duration::from_secs(SETTINGS.block_notifier.ping_interval);
+        let subscribe_interval = Duration::from_secs(SETTINGS.block_notifier.subscribe_interval);
+        Ok(ElectrumBlockNotifier {
+            client,
+            retry_interval,
+            max_retry_attempts,
+            ping_interval,
+            subscribe_interval,
+        })
+    }
+
+    /// The notify loop that handles block header notifications and sends them to the receiver.
+    async fn notify_loop(
+        client: Arc<Client>,
+        retry_interval: Duration,
+        max_retry_attempts: u32,
+        ping_interval: Duration,
+        subscribe_interval: Duration,
+        sender: Sender<Result<BlockHash, ElectrumError>>,
+    ) {
+        let mut retry_attempts = 0;
+        let mut ping_interval = interval(ping_interval);
+        let mut subscribe_interval = interval(subscribe_interval);
+        let mut continue_loop = true;
+
+        while continue_loop {
+            tokio::select! {
+                // Concurrent keep-alive check
+                _ = ping_interval.tick() => {
+                    if client.ping().is_err() {
+                        error!("Ping failed, attempting to reconnect...");
+                        retry_attempts += 1;
+                        if retry_attempts >= max_retry_attempts {
+                            error!("Max retry attempts reached, giving up.");
+                            continue_loop = false;
+                        }
+                    } else {
+                        retry_attempts = 0; // Reset retry attempts on successful ping
+                    }
+                }
+                // Periodic block header subscription
+                _ = subscribe_interval.tick() => {
+                    match client.block_headers_subscribe() {
```

Comment:
I suspect that this is a blocking call, which means that it should be executed from a separate task using `tokio::spawn_blocking`. I'm also curious how this call works, can we get the same block hash on multiple calls or is it somehow guaranteed to be incremental and how does that work? I'm looking at the `electrum-client` docs (at the [ElectrumAPI](https://docs.rs/electrum-client/0.20.0/electrum_client/trait.ElectrumApi.html#method.block_headers_subscribe) trait) but I find it hard to understand the mechanics of this call.